### 1. The probe leaks linker errors to the terminal

`$(shell ...)` captures stdout only; stderr goes straight to make's stderr. So on exactly the systems this PR targets (glibc ≤ 2.33), *every* invocation — `make`, `make clean`, `make -n` — prints this before anything happens:

```
/usr/bin/ld: /tmp/ccUVYc9i.o: in function `main':
<stdin>:(.text+0x9): undefined reference to `dlopen'
collect2: error: ld returned 1 exit status
```

Adding `2>/dev/null` inside the `$(shell ...)` fixes it.

### 2. Nothing checks that `-ldl` exists

The fallback is unconditional, so on a toolchain that has neither `dlopen()` in libc nor a `libdl` (some musl configurations, a trimmed sysroot) the build now fails with `cannot find -ldl`, which is a less informative error than the original undefined reference. Worth probing `-ldl` before adopting it, and falling back to adding nothing.

### 3. The probe uses `$(CC)` but the link is done by `$(CXX)`

`make CXX=g++-13`, or a cross-build that only sets `CXX`, will probe with the default `cc` and answer for the wrong toolchain. Probing with `$(CXX)` matches what actually performs the link. (`-m32` from the 32-bit path isn't forwarded either, so on a multilib host the probe answers for the wrong ABI — lower priority.)

### 4. `LDFLAGS +=` would avoid the new variable and the recipe change

The Makefile already accumulates link libraries in `LDFLAGS` (`-lcudart`, `-lshell32`, `-ldensity_rs`), and `$(LDFLAGS)` is already positioned after `$^` on the link line. So `LDFLAGS += $(LIBDL)` in the libbsc and nvcomp blocks gets the same result without introducing `LDFLAGS_LIBDL` and without touching the `lzbench:` recipe. (`-ldl` appearing twice when both are enabled is harmless.)

### 5. Comment wording

> very old versions of GNU Make fail if a # is included in the `$(shell ...)` function.

This understates it — the escape is required on Make 4.3 as well (I hit it on 4.3). Documenting the actual constraint would help the next reader keep the indirection intact.

### Suggested combination of the above

```make
    # some compressors need dlopen(), which is part of -lc since glibc 2.34, but
    # requires linking explicitly with -ldl on older glibc and other libc implementations.
    # NOTE: a literal '#' must be written as '\#', and GNU Make only unescapes it in a
    # variable assignment -- not inside $(shell ...) -- hence the separate variable.
    LIBDL_TEST_SRC := \#include <dlfcn.h>\nint main(){dlopen(0,0);return 0;}\n
    LIBDL_TEST_CMD  = printf '$(LIBDL_TEST_SRC)' | $(CXX) -x c++ - -o /dev/null
    ifeq ($(shell $(LIBDL_TEST_CMD) 2>/dev/null && echo ok),ok)
        # dlopen() is provided by libc, nothing to add
    else ifeq ($(shell $(LIBDL_TEST_CMD) -ldl 2>/dev/null && echo ok),ok)
        LIBDL := -ldl
    endif
```

then `LDFLAGS += $(LIBDL)` in the two blocks and no change to the link line.

I tested this on GNU Make 4.3 / glibc 2.39 across three cases — symbol in libc, symbol only in the extra library, extra library absent (the latter two simulated with `zlibVersion`/`-lz` and a nonexistent library, since everything resolves from libc on 2.39). It yields `[]`, `[-ldl]`, `[]` respectively, with no stray output in any of them. The second `ifeq` is only expanded when the first fails, so there is at most one extra compiler invocation and only on the systems that need it.

--- a/Makefile
+++ b/Makefile
@@ -80,6 +80,12 @@ else
         MOREFLAGS += -mno-strict-align
     endif
 
+    # some compressors use dlopen(), which requires linking with -ldl on glibc
+    # 2.33 and older, and other libc libraries.
+    # GNU Make 3.8.x fails to parse \# inside the $(shell ...) function.
+    LIBDL_TEST_SRC := \#include <dlfcn.h>\nint main(){dlopen(0,0);return 0;}\n
+    LIBDL := $(shell printf '${LIBDL_TEST_SRC}' | $(CXX) -x c - -o /dev/null 2>/dev/null && echo "" || echo "-ldl")
+
     # detect MacOS
     detected_OS := $(shell uname -s)
     ifeq ($(detected_OS), Darwin)
@@ -939,6 +945,8 @@ else
     BSC_CXX_FILES += bwt/libbsc/libbsc/lzp/lzp.o
     BSC_CXX_FILES += bwt/libbsc/libbsc/platform/platform.o
     BSC_CXX_FILES += bwt/libbsc/libbsc/st/st.o
+
+    LDFLAGS_LIBDL  = $(LIBDL)
 endif
 
 
@@ -1097,6 +1105,7 @@ ifeq "$(ENABLE_CUDA)" "1"
     NVCOMP_CU_SRC  = $(wildcard misc/nvcomp/src/*.cu misc/nvcomp/src/lowlevel/*.cu)
     NVCOMP_CU_OBJ  = $(NVCOMP_CU_SRC:%=%.o)
     NVCOMP_FILES   = $(NVCOMP_CU_OBJ) $(NVCOMP_CPP_OBJ)
+    LDFLAGS_LIBDL  = $(LIBDL)
   endif
 
   ifneq "$(DONT_BUILD_BSC)" "1"
@@ -1110,7 +1119,7 @@ endif # ifeq "$(ENABLE_CUDA)"
 MKDIR = mkdir -p
 
 lzbench: $(BUGGY_C_FILES) $(BUGGY_CC_FILES) $(BUGGY_CXX_FILES) $(ACEAPEX_FILES) $(BSC_C_FILES) $(BSC_CXX_FILES) $(BSC_CUDA_FILES) $(ACEAPEX_CUDA_FILES) $(BZIP2_FILES) $(BZIP3_FILES) $(CSC_FILES) $(KANZI_FILES) $(FASTLZMA2_OBJ) $(ZSTD_FILES) $(LZSSE_FILES) $(LZFSE_FILES) $(XZ_FILES) $(LIBLZG_FILES) $(BRIEFLZ_FILES) $(LZF_FILES) $(BROTLI_FILES) $(LZMA_FILES) $(ZLING_FILES) $(QUICKLZ_FILES) $(OPENZL_C_FILES) $(OPENZL_S_FILES) $(SNAPPY_FILES) $(ZLIB_FILES) $(ZLIB_NG_FILES) $(LZHAM_FILES) $(LZO_FILES) $(UCL_FILES) $(LZ4_FILES) $(LIZARD_FILES) $(LIBDEFLATE_FILES) $(ZXC_FILES) $(MISA77_FILES) $(MISC_FILES) $(NVCOMP_FILES) $(PPMD_FILES) $(BENCH_FILES) $(SKIM_FILE)
-	$(CXX) $^ -o $@ $(LDFLAGS)
+	$(CXX) $^ -o $@ $(LDFLAGS) $(LDFLAGS_LIBDL)
 	@echo Linked GCC_VERSION=$(GCC_VERSION) CLANG_VERSION=$(CLANG_VERSION) COMPILER=$(COMPILER)
 
 $(BENCH_MAIN): bench/lzbench.cpp bench/lzbench.h bench/threadpool.h bench/codecs.h DENSITY_LIB

